win,tty: improve SIGWINCH support

bzoz opened the pull request:
Add SetWinEventHook for EVENT_CONSOLE_LAYOUT for better detection of
console resize events.

Ref: https://github.com/nodejs/node/issues/13197

bzoz commented:
CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/318/

saghul commented on src/win/tty.c:
```diff
@@ -176,18 +192,22 @@ int uv_tty_init(uv_loop_t* loop, uv_tty_t* tty, uv_file fd, int readable) {
+    /* Store the global tty output handle. This handle is used by TTY read */
+    /* streams to update the virtual window when a CONSOLE_BUFFER_SIZE_EVENT */
+    /* is received. */
+    uv_tty_console_handle = handle;
```
we need to hold the uv_tty_output_lock at least here, this is not thread safe.

bzoz commented:
@saghul updated (and rebased), PTAL

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/354/

saghul commented on src/win/tty.c:
```diff
@@ -112,14 +112,30 @@ static int uv_tty_virtual_offset = -1;
+ * handle signalling SIGWINCH
+ */
+
+static HANDLE uv_tty_console_handle = INVALID_HANDLE_VALUE;
```
can you prefix these with `uv__` instead? I know src/win is not very consistent in this regard, but let's try to be, for new code.

saghul commented on src/win/tty.c:
```diff
@@ -2279,3 +2275,52 @@ static void uv__determine_vterm_state(HANDLE handle) {
+                       WINEVENT_OUTOFCONTEXT))
+    uv_fatal_error(GetLastError(), "SetWinEventHook");
+
+  while (GetMessage(&msg, NULL, 0, 0)) {
```
how is this intettupted?

saghul replied:
ping

bzoz replied:
I've described it in https://github.com/libuv/libuv/pull/1408#issuecomment-320648292

saghul commented on src/win/tty.c:
```diff
@@ -2279,3 +2275,52 @@ static void uv__determine_vterm_state(HANDLE handle) {
+  height = screen_buffer_info.srWindow.Bottom - screen_buffer_info.srWindow.Top + 1;
+
+  if (width != uv_tty_console_width || height != uv_tty_console_height) {
+    uv_tty_console_width = width;
```
this needs to hold the lock

saghul commented on src/win/tty.c:
```diff
@@ -2279,3 +2275,52 @@ static void uv__determine_vterm_state(HANDLE handle) {
+  if (!GetConsoleScreenBufferInfo(uv_tty_console_handle, &screen_buffer_info))
+    uv_fatal_error(GetLastError(), "GetConsoleScreenBufferInfo");
+
+  uv_tty_console_width = screen_buffer_info.dwSize.X;
```
this needs to hold the lock

bzoz commented:
@saghul I've updated the change, PTAL.

I've moved the `uv__tty_console_resize_message_loop_thread` start to `uv_console_init`. This way `SIGWINCH` signals can be received before any call to `uv_tty_init`.

The loop in `uv__tty_console_resize_message_loop_thread` will never be interrupted - it will spin as long as the app is alive. The `uv__tty_console_resize_event` callback will be run in the `uv__tty_console_resize_message_loop_thread` thread. Since there is only one such thread no synchronization is needed.

CI: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/364/

bzoz commented:
CI (with correct branch name): https://ci.nodejs.org/view/libuv/job/libuv-test-commit/368/

bzoz commented:
Ping?

saghul commented on src/win/tty.c:
```diff
@@ -2279,3 +2279,52 @@ static void uv__determine_vterm_state(HANDLE handle) {
+  MSG msg;
+
+  if (!GetConsoleScreenBufferInfo(uv__tty_console_handle, &sb_info))
+    uv_fatal_error(GetLastError(), "GetConsoleScreenBufferInfo");
```
Is this really fatal? Can't we live if it fails?

bzoz replied:
I guess we can, but there is no way to signal the user that this failed, and that SIGWINCH will not work.

saghul replied:
I know, but so what? it fails, we move on. I think it's far better than crashing!

saghul commented on src/win/tty.c:
```diff
@@ -2279,3 +2279,52 @@ static void uv__determine_vterm_state(HANDLE handle) {
+                       0,
+                       0,
+                       WINEVENT_OUTOFCONTEXT))
+    uv_fatal_error(GetLastError(), "SetWinEventHook");
```
ditto

saghul commented on src/win/tty.c:
```diff
@@ -2279,3 +2279,52 @@ static void uv__determine_vterm_state(HANDLE handle) {
+  int width, height;
+
+  if (!GetConsoleScreenBufferInfo(uv__tty_console_handle, &sb_info))
+    uv_fatal_error(GetLastError(), "GetConsoleScreenBufferInfo");
```
ditto

refack commented:
[biased scepticism, but I'm open to be persuaded]
I'm not sure this is worth it 🤷‍♂️  upstream we just added a disclaimer https://github.com/nodejs/node/commit/ff07eaa140e54745a5298f4c929936cfe055ed3c

Do we know of any interesting uses for this event? (honest question)
Anyone that is doing GUI-like console apps? Because very few Windows console apps react to a window change, they keep assuming the initial size and just look horrible:

![screenshot- 21 -console-rezise](https://user-images.githubusercontent.com/96947/29458939-803d769c-83ef-11e7-86e2-ef16c6b13d12.png)

saghul commented:
I hate that broken behavior. If we can do better, why not? Also, Node is not the entire universe other projects are not going to read those disclaimers.

bzoz commented:
@saghul I've removed the `uv_fatal_error` calls, PTAL

bzoz commented:
@saghul, @refack PTAL

saghul commented on src/win/tty.c:
```diff
@@ -139,8 +155,23 @@ static uv_vtermstate_t uv__vterm_state = UV_UNCHECKED;
+                          NULL,
+                          WT_EXECUTELONGFUNCTION);
+    if (r == 0)
+      abort();
```
i don't think this should be fatal

saghul commented on src/win/tty.c:
```diff
@@ -139,8 +155,23 @@ static uv_vtermstate_t uv__vterm_state = UV_UNCHECKED;
 static void uv__determine_vterm_state(HANDLE handle);
 
 void uv_console_init(void) {
+  int r;
```
you can declare r inside the if where it's used.

bzoz commented:
Updated, PTAL

bzoz commented:
Ping?

bzoz commented:
CI after rebase: https://ci.nodejs.org/view/libuv/job/libuv-test-commit/460/

bzoz commented:
Unrelated failures, rerun Windows CI: https://ci.nodejs.org/job/libuv-test-commit-windows/548/.

bzoz commented:
Landed in https://github.com/libuv/libuv/commit/6ad1e81547c83f9fe7c5a806b6b282642e8d3bca
